arraycontext/impl/pytato.py

```diff
@@ -0,0 +1,536 @@
+
+        return self._new_like(ary, _ones_like)
+
+    def maximum(self, x, y):
+        import pytato as pt
+        return rec_multimap_array_container(pt.maximum, x, y)
+
+    def minimum(self, x, y):
+        import pytato as pt
+        return rec_multimap_array_container(pt.minimum, x, y)
+
+    def where(self, criterion, then, else_):
+        import pytato as pt
+        return rec_multimap_array_container(pt.where, criterion, then, else_)
+
+    def sum(self, a, dtype=None):
+        import pytato as pt
+        if dtype not in [a.dtype, None]:
+            raise NotImplementedError
+        return pt.sum(a)
+
+    def min(self, a):
+        import pytato as pt
+        return pt.amin(a)
+
+    def max(self, a):
+        import pytato as pt
+        return pt.amax(a)
+
+    def stack(self, arrays, axis=0):
+        import pytato as pt
+        return rec_multimap_array_container(pt.stack, arrays, axis)
+
+    # {{{ relational operators
+
+    def equal(self, x, y):
+        import pytato as pt
+        return rec_multimap_array_container(pt.equal, x, y)
+
+    def not_equal(self, x, y):
+        import pytato as pt
+        return rec_multimap_array_container(pt.not_equal, x, y)
+
+    def greater(self, x, y):
+        import pytato as pt
+        return rec_multimap_array_container(pt.greater, x, y)
+
+    def greater_equal(self, x, y):
+        import pytato as pt
+        return rec_multimap_array_container(pt.greater_equal, x, y)
+
+    def less(self, x, y):
+        import pytato as pt
+        return rec_multimap_array_container(pt.less, x, y)
+
+    def less_equal(self, x, y):
+        import pytato as pt
+        return rec_multimap_array_container(pt.less_equal, x, y)
+
+    def conj(self, x):
+        import pytato as pt
+        return rec_multimap_array_container(pt.conj, x)
+
+    def arctan2(self, y, x):
+        import pytato as pt
+        return rec_multimap_array_container(pt.arctan2, y, x)
+
+    def ravel(self, a, order="C"):
+        import pytato as pt
+
+        def _rec_ravel(a):
+            if order in "FC":
+                return pt.reshape(a, (-1,), order=order)
+            elif order in "AK":
+                # memory layout is assumed to be "C"
+                return pt.reshape(a, (-1,), order="C")
+            else:
+                raise ValueError("`order` can be one of 'F', 'C', 'A' or 'K'. "
+                                 f"(got {order})")
+
+        return rec_map_array_container(_rec_ravel, a)
+
+    # }}}
+
+
+class AbstractInputDescriptor:
+    def __eq__(self, other):
+        raise NotImplementedError
+
+    def __hash__(self):
+        raise NotImplementedError
+
+
+@dataclass(frozen=True, eq=True)
+class ScalarInputDescriptor(AbstractInputDescriptor):
+    dtype: np.dtype
+
+
+@dataclass(frozen=True, eq=True)
+class ArrayContainerInputDescriptor(AbstractInputDescriptor):
+    id_to_ary_descr: "PMap[Tuple[Union[str, int], ...], Tuple[np.dtype, \
```

Comment:
Type variable, as above.

Reply:
Followed `serialize_array_container` and kept type of keys to be `Any`.